masl/masl.go

```diff
@@ -0,0 +1,421 @@
+	if err := drisl.Unmarshal(data, &d.data); err != nil {
+		return err
+	}
+	// Validate the document after unmarshaling
+	if err := d.validate(); err != nil {
+		return fmt.Errorf("invalid document after unmarshal: %w", err)
+	}
+	return nil
+}
+
+// GetResourceFromBundle retrieves a resource from the bundle by path
+func (d Document) GetResourceFromBundle(name string) (*Resource, error) {
+	// Check if this is a single-mode document (has src field)
+	if _, hasSrc := d.data["src"]; hasSrc {
+		return nil, fmt.Errorf("cannot call GetResourceFromBundle on single-mode document (use GetResource instead)")
+	}
+
+	resources, ok := d.data["resources"]
+	if !ok {
+		return nil, fmt.Errorf("document has no resources bundle")
+	}
+
+	resourcesMap, ok := resources.(map[string]any)
+	if !ok {
+		return nil, fmt.Errorf("resources field is not a map")
+	}
+
+	resourceData, ok := resourcesMap[name]
+	if !ok {
+		return nil, fmt.Errorf("resource %q not found", name)
+	}
+
+	resourceMap, ok := resourceData.(map[string]any)
+	if !ok {
+		return nil, fmt.Errorf("resource data is not a map")
+	}
+
+	return &Resource{data: resourceMap}, nil
+}
+
+// GetResource retrieves the single resource (for single-mode documents)
+func (d Document) GetResource() (*Resource, error) {
+	// Check if this is a bundle-mode document (has resources field)
+	if _, hasResources := d.data["resources"]; hasResources {
+		return nil, fmt.Errorf("cannot call GetResource on bundle-mode document (use GetResourceFromBundle instead)")
+	}
+
+	_, ok := d.data["src"]
+	if !ok {
+		return nil, fmt.Errorf("document has no src field")
+	}
+
+	return &Resource{data: d.data}, nil
+}
+
+// AddResourceToBundle adds a resource to the document's bundle
+func (d *Document) AddResourceToBundle(resource Resource, name string) error {
+	if d.data == nil {
+		d.data = make(map[string]any)
+	}
+
+	resources, ok := d.data["resources"]
+	if !ok {
+		resources = make(map[string]any)
+		d.data["resources"] = resources
+	}
+
+	resourcesMap, ok := resources.(map[string]any)
+	if !ok {
+		return fmt.Errorf("resources field is not a map")
+	}
+
+	resourcesMap[name] = resource.data
+	return nil
+}
+
+// SetSrc sets the src field for single-mode documents
+func (d *Document) SetSrc(c cid.Cid) {
+	if d.data == nil {
+		d.data = make(map[string]any)
+	}
+	d.data["src"] = c
+}
+
+// SetAttribute sets an arbitrary attribute on the document
+func (d *Document) SetAttribute(name string, value any) {
+	if d.data == nil {
+		d.data = make(map[string]any)
+	}
+	d.data[name] = value
+}
+
+// GetAttribute gets an arbitrary attribute from the document
+func (d Document) GetAttribute(name string) (any, bool) {
+	if d.data == nil {
+		return nil, false
+	}
+	value, ok := d.data[name]
+	return value, ok
+}
+
+// GetCID retrieves the CID from the resource's src field
+func (r Resource) GetCID() cid.Cid {
+	src, ok := r.data["src"]
+	if !ok {
+		return cid.Cid{}
+	}
+
+	// Handle CID stored as a link object
```

Comment:
Representations like `{"/": "bafy..."}` are how CIDs are stored in dag-json, but aren't valid in CBOR and shouldn't be handled.